### ISSUE TITLE
are there plans to generate puppet/master/remote/remotesync etc decorators?

### ISSUE THREAD

**Issue by MichaelBelousov** [GodotExplorer/ECMAScript#31]
are there plans to generate puppet/master/remote/remotesync etc decorators?

Is there any reason they haven't been done yet? Couldn't it be done with a call to rpc_config at least in the short term?

**Comment by MichaelBelousov** [comment]
I added a few decorators to do the same thing using rpc_config in my code for now, I'll edit this once I've tested them.

**Comment by MichaelBelousov** [comment]
looks like I'm definitely missing something :), my property decorators are being ran in such a way that they crash the system, I have some idea but I'll just wait for a response, and manually call rpc_config in my constructor for now.

**Comment by Geequlim** [comment]
I never using this kind of features in Godot.
I'd like to advice never touching theme as they are not battle tested features.
Fell free to implement them in script side. And pin me if anything is blocking you implement these features in the binding side.

**Comment by MichaelBelousov** [comment]
I'll look into them more, I was getting some issues on the binding side but I can look into that too

**Comment by MichaelBelousov** [comment]
What are the ramifications of this implementation? 

```ts
export function makeRsetModeConfigDecorator<T extends godot.Node, Key extends string>(mode: godot.MultiplayerAPI.RPCMode) {
	// target is prototype of property class
	return function (target: T, property: Key) {
		const originalReady = target._ready;
		target._ready = function(...args: any[]) {
			originalReady.call(this, ...args)
			this.rset_config(property, mode)
		};
	}
}

export const puppet = makeRsetModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_PUPPET);
```

Manipulating the bound `_ready` method is a bit problematic, but I'm not sure how else to inject property modifying code on object creation, and I couldn't do it with the constructor that existed on the prototype. Any ideas? I'm also knew to using decorators in typescript (I've always wanted to but it's not popular in the communities in which I work that use typescript).

This implementation does seem to work, but it's probably better to set the rset config on object construction, not tree _ready. Even better would be to figure out how it's done in gdscript and do it there so I might want to look into that. This might work for a first-pass but I'm not sure I like it. Maybe once I've done some research this implementation won't seem so bad.

**Comment by MichaelBelousov** [comment]
So indeed I can't really alter the constructor from a property decorator since that doesn't let me generate/replace the class. I could however use metadata to do something in the bindings probably.
https://stackoverflow.com/questions/9267157/why-is-it-impossible-to-change-constructor-function-from-prototype

The above solutions works for now, but I plan on contributing a better one eventually, once I start really battle-testing this one.

**Comment by MichaelBelousov** [comment]
I see hints that there may be a convention of adding `$onready:` prefixed properties that the bindings automatically execute on ready. I'll look into this later and hopefully contribute a PR. Before I do so, I am rewriting a codebase I have into ECMAScript so I have a live application to test with.

**Comment by MichaelBelousov** [comment]
I believe I have a decent solution, I'll make a PR soon, after more testing.

**Comment by MichaelBelousov** [comment]
just an update, here's the temporary solution, I'll make a PR once I have finished converting my codebase to ecmascript without any bugs, and have tested a live multiplayer scenario.

```ts
function makeRpcModeConfigDecorator<T extends godot.Node>(mode: godot.MultiplayerAPI.RPCMode) {
  return function (targetPrototype: T, property: string, descriptor?: PropertyDescriptor) {
    const originalReady = targetPrototype._ready
    Object.defineProperty(targetPrototype, property, {
      value: function (this: godot.Node) {
        this.rpc_config(property, mode)
        originalReady.call(this)
      },
    })
    return descriptor
  }
}

export function makeRsetModeConfigDecorator<T extends godot.Node, Key extends string>(
  mode: godot.MultiplayerAPI.RPCMode
) {
  return function (target: T, property: Key) {
    const originalReady = target._ready
    target._ready = function (this: godot.Node, ...args: any[]) {
      this.rset_config(property, mode)
      return originalReady.call(this, ...args)
    }
  }
}

export const remote = makeRpcModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_REMOTE)
export const master = makeRpcModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_MASTER)
export const puppet = makeRsetModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_PUPPET)
export const slave = makeRpcModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_SLAVE)
export const remote_sync = makeRpcModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_REMOTESYNC)
export const sync = makeRpcModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_SYNC)
export const master_sync = makeRsetModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_MASTERSYNC)
export const puppet_sync = makeRsetModeConfigDecorator(godot.MultiplayerAPI.RPC_MODE_PUPPETSYNC)

```

Just plop that in decorators.ts.

**Comment by Geequlim** [comment]
Are you tested them in your project ?
Is fine to add it to the decorator.ts if you really want these features.

**Comment by MichaelBelousov** [comment]
I definitely will as soon as I test them more. I tested them minimally a while ago. I can do some testing again tomorrow and submit a PR.

**Comment by MichaelBelousov** [comment]
just an update, I delayed because of the codebase I'm using is still being converted somewhat, and I noticed there are some settings I missed. I can definitely do it this weekend unless someone beats me to it.

**Comment by MichaelBelousov** [comment]
PR is up, everything looks good, solved some gotchas, still would be nice if someone else tries it (or not and I'll fix issues found later). A test project is attached to the PR thread

**Comment by MichaelBelousov** [comment]
Resolved by #53
